src/draive/splitters/basic.py

```diff
@@ -33,97 +44,91 @@ def split_text(
+    for index, splitter in enumerate(splitters):
+        parts: list[str] = text.split(splitter)
+        # if splitting has done nothing retry using next one
+        if len(parts) == 1:
+            continue
+        else:
+            # used_splitter, fallback_splitters, parts
+            return (splitter, list(splitters[index + 1 :]), parts)
     # if splitting has still done nothing then fail
-    if len(parts) == 1:
-        raise ValueError("Failed to properly split text with provided separators")
-
-    return (splitter, parts)
+    raise ValueError("Failed to properly split text with provided separators")
 
 
-def _merge(  # noqa: C901
+def _merge(  # noqa: C901, PLR0912, PLR0913
     parts: list[str],
     part_size: int,
     count_size: Callable[[str], int],
     splitter: str,
+    fallback_splitters: list[str],
     part_overlap_size: int | None,
 ) -> list[str]:
     result: list[str] = []
-    accumulator: list[str] = []
-    accumulator_size: int = 0
+    accumulator: str = ""
+    overlap_accumulator: list[str] = []
     # iterate over splitted pats
-    for part in parts:
-        temp_size: int = count_size(part)
+    last_part_index: int = len(parts) - 1
+    for index, part in enumerate(parts):
+        current_part: str
+        # Add the separator back if it's not the last part
+        if index < last_part_index:
+            current_part = part + splitter
+        else:
+            current_part = part
+
+        temp_size: int = count_size(accumulator + current_part)
+        current_part_size: int = count_size(current_part)
         # check if can add to previous part
-        if accumulator_size + temp_size < part_size:
-            accumulator.append(part)
-            accumulator_size += temp_size
+        if temp_size <= part_size:
+            accumulator += current_part
+            overlap_accumulator.append(current_part)
         # check if part is not too big on its own
-        elif temp_size > part_size:
-            chunk: str = splitter.join(accumulator).strip()
+        elif current_part_size > part_size:
+            chunk: str = accumulator.strip()
             if chunk:
                 result.append(chunk)
             # do special splitting if it is too big indeed
-            # force overlap and split on newlines and spaces
             result.extend(
                 split_text(
-                    text=part,
+                    text=current_part,
                     part_size=part_size,
-                    separators=("\n", " "),
-                    part_overlap_size=part_overlap_size
-                    or int(part_size * 0.2),  # if there was no overlap force at least 20%
+                    separators=fallback_splitters,
+                    part_overlap_size=part_overlap_size,
                     count_size=count_size,
                 ),
             )
-            accumulator = []
-            accumulator_size = 0
+            accumulator = ""
+            overlap_accumulator = []
```

Comment:
there is unhandled edge case (I believe it was not handled previously as well) - if next part is also too big we are missing the overlap 🤔

Reply:
discussed on call, overlap is when parts are smaller than the given part_size